src/System.IO.Pipelines.Compression/CompressionPipelineExtensions.cs

```diff
@@ -121,33 +121,20 @@ public WritableDeflateTransform(CompressionLevel compressionLevel, int bits)
 
                     unsafe
                     {
-                        // TODO: Pin pointer if not pinned
-                        void* inPointer;
-                        if (memory.TryGetPointer(out inPointer))
-                        {
-                            _deflater.SetInput((IntPtr)inPointer, memory.Length);
-                        }
-                        else
-                        {
-                            throw new InvalidOperationException("Pointer needs to be pinned");
-                        }
+                        var handle = memory.GetPinnedMemoryHandle();
+                        _deflater.SetInput((IntPtr)handle.PinnedPointer, memory.Length);
+                        handle.Free();
```

Comment:
Are you sure that the deflater does not need this pinned longterm?

If it does not need to be pinned longterm, this should again use fixed + DangerousGetPinnableReference